sndiokeys.1
.\"	$OpenBSD$
.\"
.\" Copyright (c) 2014-2021 Alexandre Ratchov <alex@caoua.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt SNDIOKEYS 1
.Os
.Sh NAME
.Nm sndiokeys
.Nd
control sndiod with hot-keys and play the keyboard bell
.Sh SYNOPSIS
.Nm sndiokeys
.Op Fl aDsv
.Op Fl b Ar [mod+...]key:control[+|-|!]
.Op Fl f Ar device
.Sh DESCRIPTION
.Nm
registers hot-keys in
.Xr X 7
to control
.Xr sndiod 8 .
It can also replace the
.Xr X 7
keyboard bell by a beep on the audio device.
.Pp
The default key-bindings are as follows:
.Pp
.Bl -tag -width "XF86_AudioLowerVolume" -offset indent -compact
.It XF86_AudioRaiseVolume
increase
.Cm output\.level
.It XF86_AudioLowerVolume
decrease
.Cm output\.level
.It XF86_AudioMute
toggle
.Cm output\.mute
.It Control+Mod1+0
cycle through
.Cm server\.device
.El
.Pp
.Nm
is meant to be started as a background process by
.Xr startx 1
or the display manager session startup scripts.
It exits when the X session terminates.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a
Play an beep on the audio device in place of the
.Xr X 7
keyboard bell.
.It Fl b Ar [modifier+...]key:control[+|-|!]
Increase, decrease or toggle
.Va control
whenever
.Va key
is pressed while
.Va modifier
keys are held down.
Key names correspond to symbols defined in
.Pa /usr/X11R6/include/X11/keysymdef.h
with the
.Cm XK_
prefix removed.
Supported modifiers are: Control, Mod1, and Mod4.
On most setups, Mod1 and Mod4 correspond to the Alt and Windows keys.
Available controls may be listed with the
.Xr sndioctl 1
utility.
.It Fl D
Daemonize.
.It Fl f Ar device
Audio device to control.
.It Fl s
Don't emit a beep when a control changes.
.It Fl v
Increase log verbosity.
.El
.Sh EXAMPLES
Increase or decrease the output level when Up or Down keys are pressed
while the Control and Mod1 keys are held down:
.Bd -literal -offset indent
$ sndiokeys -b Control+Mod1+Up:output.level+ \\
            -b Control+Mod1+Down:output.level-
.Ed
.Sh SEE ALSO
.Xr sndioctl 1 ,
.Xr startx 1 ,
.Xr sndio 7 ,
.Xr X 7 ,
.Xr sndiod 8
.Sh BUGS
The volume keys of multimedia keyboards may already control the output level
using another mechanism.
If so, using the same keys in
.Nm
may cause unpredictable output level changes.


sndiokeys.c
/*	$OpenBSD$	*/
/*
 * Copyright (c) 2014-2021 Alexandre Ratchov <alex@caoua.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#include <poll.h>
#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <signal.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sndio.h>
#include <X11/Xlib.h>
#include <X11/Xproto.h>
#include <X11/Xutil.h>
#include <X11/keysym.h>
#include <X11/XF86keysym.h>
#include <X11/XKBlib.h>

/*
 * The mask of modifiers supported for key-bindings.
 */
#define MODMASK		(ControlMask | Mod1Mask | Mod4Mask)

/*
 * Number of level steps between 0 and 1
 */
#define NSTEP		20

/*
 * Max fds we poll
 */
#define MAXFDS		64

/*
 * Keyboard bell parameters
 */
#define BELL_RATE	48000
#define BELL_LEN	(BELL_RATE / 20)
#define BELL_PERIOD	(BELL_RATE / 880)
#define BELL_AMP	(INT16_MAX / 32)

#define logx(n, fmt, ...)						\
	do {								\
		if (verbose >= n)					\
			fprintf(stderr, fmt "\n", ## __VA_ARGS__);	\
	} while (0)

struct modname {
	unsigned int modmask;
	char *name;
} modname_tab[] = {
	{ControlMask, "Control"},
	{Mod1Mask, "Mod1"},
	{Mod4Mask, "Mod4"},
	{0, NULL}
};

struct ctl {
	struct ctl *next;
	struct sioctl_desc desc;
	int val;
} *ctl_list;

struct key {
	struct key *next;
	unsigned int modmask;
	KeySym sym;
	KeyCode code;
	KeySym *map;
	char *name;
	char *func;
	int dir;
} *key_list;

Display	*dpy;
int (*error_handler_xlib)(Display *, XErrorEvent *);
KeySym error_keysym;

char *dev_name;
struct sio_hdl *beep_hdl;
int beep_maxfds;
struct sioctl_hdl *ctl_hdl;
int ctl_maxfds;
int maxfds;

int verbose;
int silent;
int beep_pending;
int audible_bell;

static void
beep_close(void)
{
	maxfds -= beep_maxfds;
	sio_close(beep_hdl);
	beep_hdl = NULL;
}

static int
beep_open(void)
{
	struct sio_par par;

	beep_hdl = sio_open(dev_name, SIO_PLAY, 0);
	if (beep_hdl == NULL) {
		logx(1, "bell: failed to open audio device");
		return 0;
	}

	sio_initpar(&par);
	par.bits = 16;
	par.rate = BELL_RATE;
	par.pchan = 1;

	if (!sio_setpar(beep_hdl, &par) || !sio_getpar(beep_hdl, &par)) {
		logx(1, "bell: failed to set parameters");
		goto err_close;
	}

	if (par.bits != 16 || par.bps != 2 || par.le != SIO_LE_NATIVE ||
	    par.pchan != 1 || par.rate != BELL_RATE) {
		logx(1, "bell: bad parameters");
		goto err_close;
	}
	beep_maxfds = sio_nfds(beep_hdl);
	if (beep_maxfds + maxfds >= MAXFDS) {
		logx(1, "%s: too many fds", dev_name);
		goto err_close;
	}
	maxfds += beep_maxfds;
	return 1;
err_close:
	sio_close(beep_hdl);
	beep_hdl = NULL;
	return 0;
}

/*
 * Play a short beep. It's used as sonic feedback and/or keyboard bell
 */
static void
beep_play(void)
{
	int16_t data[BELL_LEN];
	int i;

	if (beep_hdl == NULL) {
		if (!beep_open())
			return;
	}
	if (!sio_start(beep_hdl)) {
		logx(1, "bell: failed to start playback");
		return;
	}
	for (i = 0; i < BELL_LEN; i++) {
		data[i] = (i % BELL_PERIOD) < (BELL_PERIOD / 2) ?
		    BELL_AMP : -BELL_AMP;
	}
	if (sio_write(beep_hdl, data, sizeof(data)) != sizeof(data)) {
		logx(1, "bell: short write");
		return;
	}
	sio_stop(beep_hdl);
}

/*
 * compare two sioctl_desc structures, used to sort ctl_list
 */
static int
cmpdesc(struct sioctl_desc *d1, struct sioctl_desc *d2)
{
	int res;

	res = strcmp(d1->group, d2->group);
	if (res != 0)
		return res;
	res = strcmp(d1->node0.name, d2->node0.name);
	if (res != 0)
		return res;
	res = d1->type - d2->type;
	if (res != 0)
		return res;
	res = strcmp(d1->func, d2->func);
	if (res != 0)
		return res;
	res = d1->node0.unit - d2->node0.unit;
	if (d1->type == SIOCTL_SEL) {
		if (res != 0)
			return res;
		res = strcmp(d1->node1.name, d2->node1.name);
		if (res != 0)
			return res;
		res = d1->node1.unit - d2->node1.unit;
	}
	return res;
}

/*
 * skip all parameters with the same group, name, and func
 */
static struct ctl *
nextctl(struct ctl *i)
{
	char *str, *group, *func;
	int unit;

	group = i->desc.group;
	func = i->desc.func;
	str = i->desc.node0.name;
	unit = i->desc.node0.unit;
	for (i = i->next; i != NULL; i = i->next) {
		if (strcmp(i->desc.group, group) != 0 ||
		    strcmp(i->desc.node0.name, str) != 0 ||
		    strcmp(i->desc.func, func) != 0 ||
		    i->desc.node0.unit != unit)
			return i;
	}
	return NULL;
}

/*
 * move to the next selector entry or retrun NULL
 */
static struct ctl *
nextent(struct ctl *i)
{
	char *str, *group, *func;
	int unit;

	group = i->desc.group;
	func = i->desc.func;
	str = i->desc.node0.name;
	unit = i->desc.node0.unit;
	for (i = i->next; i != NULL; i = i->next) {
		if (strcmp(i->desc.group, group) != 0 ||
		    strcmp(i->desc.node0.name, str) != 0 ||
		    strcmp(i->desc.func, func) != 0)
			return NULL;
		if (i->desc.node0.unit == unit)
			return i;
	}
	return NULL;
}

/*
 * sndio call-back for added/removed controls
 */
static void
ondesc(void *unused, struct sioctl_desc *desc, int val)
{
	struct ctl *i, **pi;

	if (desc == NULL)
		return;

	for (pi = &ctl_list; (i = *pi) != NULL; pi = &i->next) {
		if (desc->addr == i->desc.addr) {
			*pi = i->next;
			free(i);
			break;
		}
	}

	switch (desc->type) {
	case SIOCTL_NUM:
	case SIOCTL_SW:
	case SIOCTL_SEL:
		break;
	default:
		return;
	}

	/*
	 * find the right position to insert the new widget
	 */
	for (pi = &ctl_list; (i = *pi) != NULL; pi = &i->next) {
		if (cmpdesc(desc, &i->desc) <= 0)
			break;
	}

	i = malloc(sizeof(struct ctl));
	if (i == NULL) {
		logx(1, "failed to allocate desc: %s", strerror(errno));
		exit(1);
	}
	i->desc = *desc;
	i->val = val;
	i->next = *pi;
	*pi = i;
}

/*
 * sndio call-back for control value/changes
 */
static void
onval(void *unused, unsigned int addr, unsigned int val)
{
	struct ctl *i, *j;

	logx(1, "onval: %d -> %d", addr, val);

	i = ctl_list;
	for (;;) {
		if (i == NULL)
			return;
		if (i->desc.addr == addr)
			break;
		i = i->next;
	}

	if (i->desc.type == SIOCTL_SEL) {
		for (j = ctl_list; j != NULL; j = j->next) {
			if (strcmp(i->desc.group, j->desc.group) != 0 ||
			    strcmp(i->desc.node0.name, j->desc.node0.name) != 0 ||
			    strcmp(i->desc.func, j->desc.func) != 0 ||
			    i->desc.node0.unit != j->desc.node0.unit)
				continue;
			j->val = (i->desc.addr == j->desc.addr);
		}
	} else
		i->val = val;

}

static int
ctl_open(void)
{
	ctl_hdl = sioctl_open(dev_name, SIOCTL_READ | SIOCTL_WRITE, 0);
	if (ctl_hdl == NULL) {
		logx(1, "%s: couldn't open audio device", dev_name);
		return 0;
	}
	sioctl_ondesc(ctl_hdl, ondesc, NULL);
	sioctl_onval(ctl_hdl, onval, NULL);

	ctl_maxfds = sioctl_nfds(ctl_hdl);
	if (ctl_maxfds + maxfds >= MAXFDS) {
		logx(1, "%s: too many fds", dev_name);
		sioctl_close(ctl_hdl);
		ctl_hdl = NULL;
		return 0;
	}
	maxfds += ctl_maxfds;
	return 1;
}

static void
ctl_close(void)
{
	struct ctl *c;

	while ((c = ctl_list) != NULL) {
		ctl_list = ctl_list->next;
		free(c);
	}
	maxfds -= ctl_maxfds;
	sioctl_close(ctl_hdl);
	ctl_hdl = NULL;
}

static void
setval_sel(struct ctl *first, int dir)
{
	struct ctl *cur, *next;

	if (dir != 0)
		return;

	/* find the current entry */

	cur = first;
	while (cur->val == 0) {
		cur = nextent(cur);
		if (cur == NULL) {
			logx(1, "%d: no current value", first->desc.addr);
			return;
		}
	}

	/* find the next entry */

	next = nextent(cur);
	if (next == NULL)
		next = first;
	if (next == cur) {
		logx(1, "%d: no next value", first->desc.addr);
		return;
	}

	logx(2, "%d -> %s", next->desc.addr, next->desc.node1.name);

	cur->val = 0;
	next->val = 1;
	sioctl_setval(ctl_hdl, next->desc.addr, 1);
	if (!silent)
		beep_pending = 1;
}

static void
setval_num(struct ctl *i, int dir)
{
	int val, incr;

	if (i->desc.maxval > 1 && dir != 0) {
		incr = ((int)i->desc.maxval + NSTEP - 1) / NSTEP;
		val = i->val + dir * incr;
		if (val < 0)
			val = 0;
		if (val > i->desc.maxval)
			val = i->desc.maxval;
		if (val == i->val)
			return;
	} else if (i->desc.maxval == 1 && dir == 0) {
		val = i->val ^ 1;
	} else
		return;

	logx(2, "%d -> %d", i->desc.addr, val);

	i->val = val;
	sioctl_setval(ctl_hdl, i->desc.addr, val);
	if (!silent)
		beep_pending = 1;
}

/*
 * change the control
 */
static void
setval(char *name, char *func, int dir)
{
	struct ctl *i;

	if (!ctl_hdl) {
		if (!ctl_open())
			return;
	}

	i = ctl_list;
	while (1) {
		if (i == NULL)
			return;
		if (i->desc.group[0] == 0 &&
		    strcmp(i->desc.node0.name, name) == 0 &&
		    strcmp(i->desc.func, func) == 0) {
			if (i->desc.type == SIOCTL_SEL)
				setval_sel(i, dir);
			else
				setval_num(i, dir);
		}
		i = nextctl(i);
	}
}

/*
 * error handler for Xlib functions. Print a meaningful
 * message for well-known errors and exit.
 */
static int
error_handler(Display *d, XErrorEvent *e)
{
	if (e->request_code == X_GrabKey && e->error_code == BadAccess) {
		logx(1, "Key \"%s\" already grabbed by another program",
		    XKeysymToString(error_keysym));
		exit(1);
	}

	return error_handler_xlib(d, e);
}

/*
 * register hot-keys
 */
static void
grab_keys(void)
{
	struct key *key;
	unsigned int i, scr, nscr;
	int nret;

	for (key = key_list; key != NULL; key = key->next) {

		key->code = XKeysymToKeycode(dpy, key->sym);
		key->map = XGetKeyboardMapping(dpy, key->code, 1, &nret);
		if (nret <= ShiftMask) {
			logx(1, "%s: couldn't get keymap for key",
			    XKeysymToString(key->sym));
			exit(1);
		}

		/*
		 * Grab the key for all the modifier combinations
		 * whose MODMASK bits match exactly. This way
		 * X sends the events regardless of the state
		 * of the other modifiers: Shift, Caps Lock,
		 * Num Lock, Scroll Lock and Mode switch.
		 */
		error_keysym = key->sym;
		nscr = ScreenCount(dpy);
		for (i = 0; i <= 0xff; i++) {
			if ((i & MODMASK) != key->modmask)
				continue;
			for (scr = 0; scr != nscr; scr++) {
				XGrabKey(dpy, key->code, i,
				    RootWindow(dpy, scr), 1,
				    GrabModeAsync, GrabModeAsync);
			}
		}
		XSync(dpy, False);
	}
}

/*
 * unregister hot-keys
 */
static void
ungrab_keys(void)
{
	struct key *key;
	unsigned int scr, nscr;

	for (key = key_list; key != NULL; key = key->next) {
		XFree(key->map);
	}

	nscr = ScreenCount(dpy);
	for (scr = 0; scr != nscr; scr++)
		XUngrabKey(dpy, AnyKey, AnyModifier, RootWindow(dpy, scr));
}

/*
 * add key binding, removing old binding for the same function
 */
static void
add_key(unsigned int modmask, KeySym sym, char *name, char *func, int dir)
{
	struct key *key, **p;

	/* delete existing bindings for the same function */
	p = &key_list;
	while ((key = *p) != NULL) {
		if (strcmp(key->name, name) == 0 &&
		    strcmp(key->func, func) == 0 && key->dir == dir) {
			*p = key->next;
			free(key);
		} else
			p = &key->next;
	}

	key = malloc(sizeof(struct key));
	if (key == NULL) {
		logx(1, "failed to allocate key: %s", strerror(errno));
		exit(1);
	}
	key->sym = sym;
	key->modmask = modmask;
	key->name = name;
	key->func = func;
	key->dir = dir;

	key->next = NULL;
	*p = key;
}

/*
 * parse key binding with this format:
 *
 *	[mod '+' mod '+' ...] key ':' name '.' func {'+' | '-' | '!'}
 */
static void
parsekey(char *str)
{
	char *p, *end, *name, *func;
	struct modname *mod;
	unsigned int modmask;
	KeySym keysym;
	int dir;

	name = strchr(str, ':');
	if (name == NULL) {
		logx(1, "%s: expected ':'", str);
		exit(1);
	}
	*name++ = 0;

	p = str;

	modmask = 0;
	while (1) {
		end = strchr(p, '+');
		if (end == NULL)
			break;
		*end = 0;

		mod = modname_tab;
		while (1) {
			if (mod->modmask == 0) {
				logx(1, "%s: bad modifier", p);
				exit(1);
			}
			if (strcmp(p, mod->name) == 0) {
				modmask |= mod->modmask;
				break;
			}
			mod++;
		}

		p = end + 1;
	}

	keysym = XStringToKeysym(p);
	if (keysym == NoSymbol) {
		logx(1, "%s: unknowm key", p);
		exit(1);
	}

	/*
	 * compat with old versions
	 */
	if (strcmp(name, "inc_level") == 0) {
		add_key(modmask, keysym, "output", "level", 1);
		return;
	} else if (strcmp(name, "dec_level") == 0) {
		add_key(modmask, keysym, "output", "level", -1);
		return;
	} else if (strcmp(name, "cycle_dev") == 0) {
		add_key(modmask, keysym, "server", "device", 0);
		return;
	}

	func = strchr(name, '.');
	if (func == NULL) {
		fprintf(stderr, "%s: expected '.'", name);
		exit(1);
	}
	*func++ = 0;

	if ((end = strchr(func, '+')) != NULL) {
		dir = 1;
	} else if ((end = strchr(func, '-')) != NULL) {
		dir = -1;
	} else if ((end = strchr(func, '!')) != NULL) {
		dir = 0;
	} else {
		fprintf(stderr, "%s: expected '+', '-' or '!'", end);
		exit(1);
	}
	*end++ = 0;

	if (*end != 0) {
		logx(1, "%s: junk at end of the argument", end);
		exit(1);
	}

	add_key(modmask, keysym, name, func, dir);
}

int
main(int argc, char **argv)
{
	int scr;
	XEvent xev;
	int c, nfds, ctl_nfds, beep_nfds;
	int background;
	struct pollfd pfds[MAXFDS];
	struct key *key;
	int xkb, xkb_ev_base, xkb_auto_controls, xkb_auto_values;

	dev_name = SIO_DEVANY;
	verbose = 1;
	background = 0;

	add_key(0, XF86XK_AudioRaiseVolume, "output", "level", 1);
	add_key(0, XF86XK_AudioLowerVolume, "output", "level", -1);
	add_key(0, XF86XK_AudioMute, "output", "mute", 0);
	add_key(ControlMask | Mod1Mask, XK_0, "server", "device", 0);

	while ((c = getopt(argc, argv, "ab:Df:m:sv")) != -1) {
		switch (c) {
		case 'a':
			audible_bell = 1;
			break;
		case 'b':
			parsekey(optarg);
			break;
		case 'D':
			background = 1;
			break;
		case 'f':
			dev_name = optarg;
			break;
		case 's':
			silent = 1;
			break;
		case 'v':
			verbose++;
			break;
		default:
			goto bad_usage;
		}
	}
	argc -= optind;
	argv += optind;

	if (argc > 0) {
	bad_usage:
		fputs("usage: sndiokeys "
		    "[-aDsv] "
		    "[-b [mod+...]key:control[+|-|!] "
		    "[-f device]\n",
		    stderr);
		exit(1);
	}

	error_handler_xlib = XSetErrorHandler(error_handler);

	dpy = XOpenDisplay(NULL);
	if (dpy == 0) {
		logx(1, "Couldn't open display");
		exit(1);
	}
	maxfds++;

	xkb = 0;
	if (audible_bell) {
		if (XkbQueryExtension(dpy, NULL, &xkb_ev_base, NULL, NULL, NULL)) {
			XkbSelectEvents(dpy, XkbUseCoreKbd,
			    XkbBellNotifyMask, XkbBellNotifyMask);
			xkb_auto_controls = XkbAudibleBellMask;
			xkb_auto_values = XkbAudibleBellMask;
			XkbSetAutoResetControls(dpy,
			    XkbAudibleBellMask, &xkb_auto_controls, &xkb_auto_values);
			XkbChangeEnabledControls(dpy, XkbUseCoreKbd,
			    XkbAudibleBellMask, 0);
			xkb = 1;
		} else
			logx(1, "Audible bell not suppored by the X server");
	}

	/* mask non-key events for each screan */
	for (scr = 0; scr != ScreenCount(dpy); scr++)
		XSelectInput(dpy, RootWindow(dpy, scr), KeyPress);

	grab_keys();

	if (background) {
		verbose = 0;
		if (daemon(0, 0) < 0) {
			logx(1, "failed to daemonize: %s", strerror(errno));
			exit(1);
		}
	}

	while (1) {
		while (XPending(dpy)) {
			XNextEvent(dpy, &xev);
			if (xev.type == MappingNotify) {
				if (xev.xmapping.request != MappingKeyboard)
					continue;
				logx(1, "keyboard remapped");
				ungrab_keys();
				grab_keys();
				continue;
			}
			if (xkb && xev.type == xkb_ev_base &&
			    ((XkbEvent *)&xev)->any.xkb_type == XkbBellNotify) {
				beep_pending = 1;
				continue;
			}
			if (xev.type != KeyPress)
				continue;
			for (key = key_list; key != NULL; key = key->next) {
				if (xev.xkey.keycode == key->code &&
				    key->map[xev.xkey.state & ShiftMask] == key->sym &&
				    key->modmask == (xev.xkey.state & MODMASK))
					setval(key->name, key->func, key->dir);
			}
		}

		/*
		 * keyboard auto-repeat may try to play beep multiple times,
		 * play it only once
		 */
		if (beep_pending) {
			beep_play();
			beep_pending = 0;
		}

		nfds = 0;
		if (ctl_hdl) {
			ctl_nfds = sioctl_pollfd(ctl_hdl, pfds + nfds, 0);
			nfds += ctl_nfds;
		}
		if (beep_hdl) {
			beep_nfds = sio_pollfd(beep_hdl, pfds + nfds, 0);
			nfds += beep_nfds;
		}
		pfds[nfds].fd = ConnectionNumber(dpy);
		pfds[nfds].events = POLLIN;
		nfds++;

		while (poll(pfds, nfds, -1) < 0 && errno == EINTR)
			; /* nothing */

		nfds = 0;
		if (ctl_hdl) {
			if (sioctl_revents(ctl_hdl, pfds + nfds) & POLLHUP) {
				logx(1, "sndio: ctl hup");
				ctl_close();
			}
			nfds += ctl_nfds;
		}
		if (beep_hdl) {
			if (sio_revents(beep_hdl, pfds + nfds) & POLLHUP) {
				logx(1, "sndio: beep hup");
				beep_close();
			}
			nfds += beep_nfds;
		}
		if (pfds[nfds].revents & POLLHUP) {
			logx(1, "x11: hup");
			break;
		}
		nfds++;
	}

	ungrab_keys();
	XCloseDisplay(dpy);
	maxfds--;

	if (ctl_hdl)
		ctl_close();

	if (beep_hdl)
		beep_close();

	while ((key = key_list) != NULL) {
		key_list = key_list->next;
		free(key);
	}

	return 0;
}
